pkg/deploy/apiserver/apiserver.go

```diff
@@ -0,0 +1,168 @@
+package apiserver
+
+import (
+	"fmt"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/apimachinery/registered"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apiserver/pkg/registry/rest"
+	genericapiserver "k8s.io/apiserver/pkg/server"
+	restclient "k8s.io/client-go/rest"
+	kapi "k8s.io/kubernetes/pkg/api"
+	kclientsetexternal "k8s.io/kubernetes/pkg/client/clientset_generated/clientset"
+	kclientsetinternal "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	kubeletclient "k8s.io/kubernetes/pkg/kubelet/client"
+
+	osclient "github.com/openshift/origin/pkg/client"
+	appsapiv1 "github.com/openshift/origin/pkg/deploy/apis/apps/v1"
+	oappsclient "github.com/openshift/origin/pkg/deploy/generated/internalclientset"
+	deployconfigetcd "github.com/openshift/origin/pkg/deploy/registry/deployconfig/etcd"
+	deploylogregistry "github.com/openshift/origin/pkg/deploy/registry/deploylog"
+	deployconfiginstantiate "github.com/openshift/origin/pkg/deploy/registry/instantiate"
+	deployrollback "github.com/openshift/origin/pkg/deploy/registry/rollback"
+)
+
+type AppsConfig struct {
+	GenericConfig *genericapiserver.Config
+
+	CoreAPIServerClientConfig *restclient.Config
+	KubeletClientConfig       *kubeletclient.KubeletClientConfig
+
+	// TODO these should all become local eventually
+	Scheme   *runtime.Scheme
+	Registry *registered.APIRegistrationManager
+	Codecs   serializer.CodecFactory
+
+	makeV1Storage sync.Once
+	v1Storage     map[string]rest.Storage
+	v1StorageErr  error
+}
+
+type AppsServer struct {
+	GenericAPIServer *genericapiserver.GenericAPIServer
+}
+
+type completedConfig struct {
+	*AppsConfig
+}
+
+// Complete fills in any fields not set that are required to have valid data. It's mutating the receiver.
+func (c *AppsConfig) Complete() completedConfig {
+	c.GenericConfig.Complete()
+
+	return completedConfig{c}
+}
+
+// SkipComplete provides a way to construct a server instance without config completion.
+func (c *AppsConfig) SkipComplete() completedConfig {
+	return completedConfig{c}
+}
+
+// New returns a new instance of AppsServer from the given config.
+func (c completedConfig) New(delegationTarget genericapiserver.DelegationTarget) (*AppsServer, error) {
+	genericServer, err := c.AppsConfig.GenericConfig.SkipComplete().New("apps.openshift.io-apiserver", delegationTarget) // completion is done in Complete, no need for a second time
```

Comment:
move the comment on top

Reply:
also `apps.openshift.io-apiserver` looks like const?

Reply:
also is `io-apiserver` some new pattern or why we can't use `apps.openshift.io/apiserver` ?